extensions-core/google-extensions/src/main/java/org/apache/druid/storage/google/GoogleTimestampVersionedDataFinder.java

```diff
@@ -0,0 +1,93 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.storage.google;
+
+import com.google.api.services.storage.model.Objects;
+import com.google.api.services.storage.model.StorageObject;
+import com.google.inject.Inject;
+import org.apache.druid.data.SearchableVersionedDataFinder;
+import org.apache.druid.data.input.impl.CloudObjectLocation;
+import org.apache.druid.java.util.common.StringUtils;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.net.URI;
+import java.util.regex.Pattern;
+
+public class GoogleTimestampVersionedDataFinder extends GoogleDataSegmentPuller
+    implements SearchableVersionedDataFinder<URI>
+{
+  private static final long MAX_LISTING_KEYS = 1000;
+
+  @Inject
+  public GoogleTimestampVersionedDataFinder(final GoogleStorage storage)
+  {
+    super(storage);
+  }
+
+  @Override
+  public URI getLatestVersion(URI descriptorBase, @Nullable Pattern pattern)
+  {
+    try {
+      long mostRecent = Long.MIN_VALUE;
+      URI latest = null;
+      final String bucket = descriptorBase.getAuthority();
+      final String prefix = StringUtils.maybeRemoveLeadingSlash(descriptorBase.getPath());
+      final Objects objects = storage.list(bucket).setPrefix(prefix).setMaxResults(MAX_LISTING_KEYS).execute();
+      for (StorageObject storageObject : objects.getItems()) {
+        if (isDirectoryPlaceholder(storageObject)) {
+          continue;
+        }
+        // remove path prefix from file name
+        final CloudObjectLocation objectLocation = new CloudObjectLocation(storageObject.getBucket(),
+            storageObject.getName()
+        );
+        final String keyString = StringUtils
+            .maybeRemoveLeadingSlash(storageObject.getName().substring(prefix.length()));
+        if (pattern != null && !pattern.matcher(keyString).matches()) {
+          continue;
+        }
+        final long latestModified = storageObject.getUpdated().getValue();
+        if (latestModified >= mostRecent) {
+          mostRecent = latestModified;
+          latest = objectLocation.toUri(GoogleStorageDruidModule.SCHEME_GS);
+        }
+      }
+      return latest;
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Similar to {@link org.apache.druid.storage.s3.ObjectSummaryIterator#isDirectoryPlaceholder}
+   * Copied to avoid creating dependency on s3 extensions
+   */
+  private static boolean isDirectoryPlaceholder(final StorageObject storageObject)
+  {
+    // Recognize "standard" directory place-holder indications
+    if (storageObject.getName().endsWith("/") && storageObject.getSize().intValue() == 0) {
+      return true;
+    }
+    // Recognize place-holder objects created by the Google Storage console or S3 Organizer Firefox extension.
+    return storageObject.getName().endsWith("_$folder$") && storageObject.getSize().intValue() == 0;
+  }
```

Comment:
there is a similar check being done in `ObjectStorageIterator.advanceStorageObject`, though this one looks like it might be better at picking up more stuff that Druid probably doesn't want to see, maybe this method should be moved to `GoogleUtils` or somewhere and shared?

Reply:
moved to `GoogleUtils`